File: 1338-reduce-array-size-to-the-half/1338-reduce-array-size-to-the-half.cpp
class Solution {
    static const int MAX = 100000;
public:
    int minSetSize(vector<int>& arr) {        
        int freq[MAX+1]{}, freqBuckets[MAX+1]{}, ans = 0, maxFreq = 0;
        for(auto num : arr) maxFreq = max(maxFreq, ++freq[num]); 
        for(int i = 0; i <= MAX; i++) freqBuckets[freq[i]]++;   
        for(int i = maxFreq, deleteRequired = size(arr) / 2; deleteRequired > 0; i--) {
            if(!freqBuckets[i]) continue; 
            int noOfFreqToDelete = min(freqBuckets[i], deleteRequired / i + (deleteRequired % i > 0)); 
            deleteRequired -= i * noOfFreqToDelete; 
            ans += noOfFreqToDelete;                             
        }   
        return ans;
    }
};

File: 1338-reduce-array-size-to-the-half/1338-reduce-array-size-to-the-half.java
class Solution {
   public int minSetSize(int[] arr) {
        int len = arr.length;
        //step1: map(num -> freq)
        HashMap<Integer, Integer> map = new HashMap<>();
        int maxFreq=0;
        for(int num : arr ){
            map.put(num, map.getOrDefault(num, 0)+1);
            maxFreq = Math.max(maxFreq, map.get(num));
        }
        //step2 buck sort: int[](freq -> NUM[num1, num2, ...])
        int[] freqArr = new int[maxFreq+1];
        for(int freq : map.values()){
            freqArr[freq]++;
        }
        //step3:
        int res=0, half=len/2, removed=0, curFreq=maxFreq;
        while(removed<half){
            while(freqArr[curFreq]==0) curFreq--;
            freqArr[curFreq]--;
            removed+=curFreq;
            res++;
        }
        return res;
    } 
}